src/invocation/ClientConnectionManager.ts

```diff
@@ -51,13 +51,16 @@ class ClientConnectionManager {
             var callback = (authenticated: boolean) => {
                 if (authenticated) {
                     result.resolve(connection);
-                    this.establishedConnections[addressIndex] = connection;
+                    this.establishedConnections[connection.address.toString()] = connection;
```

Comment:
Yes, if user gives `localhost` as server address we connect to localhost but then we refer the same connection with its IP address returned by server after authentication. It causes complexity in listener events. One example is heartbeat stop. Owner connection stops heartbeating but later connection manager cannot find right listener to notify because the connection is referred with name `localhost` instead of its IP.